src/bridge/cockpitdbusmachines.c

```diff
@@ -375,6 +385,69 @@ on_machines_changed (GFileMonitor *monitor,
   g_free (path);
 }
 
+static void
+migrate_var_config (void)
+{
+  const gchar *var_path = "/var/lib/cockpit/machines.json";
+  GError *error = NULL;
+
+  /* TOCTOU, but if we really miss this, we'll migrate it the next time */
+  if (!g_file_test (var_path, G_FILE_TEST_IS_REGULAR))
+    {
+      g_debug ("%s does not exist, nothing to migrate", var_path);
+      return;
+    }
+
+  /* the directory should already exist (shipped by the package), but let's make sure */
+  if (g_mkdir_with_parents (machines_json_dir (), 0755) < 0)
+    {
+      g_warning ("failed to create %s, Cockpit will not work properly: %m", machines_json_dir ());
+      return;
+    }
+
+  /* common case is to move it to 99-webui.json */
+  gchar *etc_path = g_build_filename (machines_json_dir (), "99-webui.json", NULL);
+  GFile *var_file = g_file_new_for_path (var_path);
+  GFile *etc_file = g_file_new_for_path (etc_path);
+  if (g_file_move (var_file, etc_file, G_FILE_COPY_NONE, NULL, NULL, NULL, &error))
+    {
+      g_info ("migrated %s to %s", var_path, etc_path);
+    }
+  else
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          GError *error2 = NULL;
+
+          /* most likely an interrupted/failed previous transition attempt;
+           * don't clobber the existing file but move it to 98-migrated.json
+           * instead */
+          g_free (etc_path);
+          g_object_unref (etc_file);
+          etc_path = g_build_filename (machines_json_dir (), "98-migrated.json", NULL);
+          etc_file = g_file_new_for_path (etc_path);
+          if (g_file_move (var_file, etc_file, G_FILE_COPY_NONE, NULL, NULL, NULL, &error2))
+            {
+              g_info ("migrated %s to %s (99-webui.json already exists)", var_path, etc_path);
+            }
+          else
+            {
+              g_warning ("moving of %s to %s failed: %s", var_path, etc_path, error2->message);
+              g_error_free (error2);
+            }
+        }
+      else /* different g_file_move() error than EXISTS */
+        {
+          g_warning ("migration of %s to %s failed: %s", var_path, etc_path, error->message);
```

Comment:
At this point we already asserted `g_access (machines_json_dir (), W_OK)` (from the call below), so this could happen if `/var/lib/cockpit/` isn't writable. So this really is a Should Not Happen™ situation, and if someone gets here they wedged up their system. So IMHO `g_warning()` is not too strong at all.